### PR TITLE
fix initial sync

### PR DESCRIPTION
**Pull request by shayzluf**
[Resolves] #3665

---

# Description
fix initial sync issue

The issue was made of few parts . 

- ```if count==0{count=1}``` was masking out many skipped slots
- step was re increasing itself on the next for loop
- subsequent requests on failure was with lower count then needed 
- if "best peer" has failures initial sync fails

### CODE REVIEW

File: beacon-chain/sync/initial-sync/round_robin.go [hunk 1]
```diff
@@ -44,7 +44,9 @@ func (s *InitialSync) roundRobinSync(genesis time.Time) error {
 	var lastEmptyRequests int
 	errChan := make(chan error)
 	// Step 1 - Sync to end of finalized epoch.
-	for s.chain.HeadSlot() < helpers.StartSlot(highestFinalizedEpoch()+1) {
+	for s.chain.HeadSlot() < helpers.StartSlot(highestFinalizedEpoch()) {
+		log.WithField("head status:", s.chain.HeadSlot()).Debugf("helpers.StartSlot(highestFinalizedEpoch()+1)-1: %v", helpers.StartSlot(highestFinalizedEpoch()))
```
Comment (terencechain) [inline beacon-chain/sync/initial-sync/round_robin.go]: i would not put this as normal debug log

File: beacon-chain/sync/initial-sync/round_robin.go [hunk 2]
```diff
@@ -70,24 +72,25 @@ func (s *InitialSync) roundRobinSync(genesis time.Time) error {
 				if ctx.Err() != nil {
 					return nil, ctx.Err()
 				}
-				start := start + uint64(i)*step
-				step := step * uint64(len(peers))
-				count := mathutil.Min(count, (helpers.StartSlot(finalizedEpoch+1)-start)/step)
+				stp := step * uint64(len(peers))
```
Comment (terencechain) [inline beacon-chain/sync/initial-sync/round_robin.go]: what are stp and str? can we use better names?

File: beacon-chain/sync/initial-sync/round_robin.go (hunk 1 above)
Comment (shayzluf) [reply beacon-chain/sync/initial-sync/round_robin.go]: removed the log

File: beacon-chain/sync/initial-sync/round_robin.go (hunk 2 above)
Comment (shayzluf) [reply beacon-chain/sync/initial-sync/round_robin.go]: just not to run over the same vars stp is for step cnt is for count ...

**Comment by terencechain** [comment]
Will defer to @prestonvanloon and @nisdas as they have the most knowledge in this subject

File: beacon-chain/sync/initial-sync/round_robin.go [hunk 3]
```diff
@@ -70,24 +67,24 @@ func (s *InitialSync) roundRobinSync(genesis time.Time) error {
...
+				str := start + uint64(i)*step
+				cnt := mathutil.Min(count, (helpers.StartSlot(finalizedEpoch+1)-str)/stp)
 				// If the count was divided by an odd number of peers, there will be some blocks
 				// missing from the first requests so we accommodate that scenario.
 				if i < remainder {
-					count++
+					cnt++
 				}
 				// asking for no blocks may cause the client to hang. This should never happen and
 				// the peer may return an error anyway, but we'll ask for at least one block.
```
Comment (prestonvanloon) [inline beacon-chain/sync/initial-sync/round_robin.go]: Please edit comment

File: beacon-chain/sync/initial-sync/round_robin.go (hunk 2 above)
Comment (prestonvanloon) [reply beacon-chain/sync/initial-sync/round_robin.go]: stp and str are not clear enough. was there a problem with the variable shadowing?

File: beacon-chain/sync/initial-sync/round_robin.go [hunk 4]
```diff
@@ -38,15 +38,12 @@ const counterSeconds = 20
 func (s *InitialSync) roundRobinSync(genesis time.Time) error {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
-
 	counter := ratecounter.NewRateCounter(counterSeconds * time.Second)
-
 	var lastEmptyRequests int
 	errChan := make(chan error)
 	// Step 1 - Sync to end of finalized epoch.
-	for s.chain.HeadSlot() < helpers.StartSlot(highestFinalizedEpoch()+1) {
```
Comment (prestonvanloon) [inline beacon-chain/sync/initial-sync/round_robin.go]: Please revert this. We want to sync to the start of the next epoch after finalized.

File: beacon-chain/sync/initial-sync/round_robin.go [hunk 5]
```diff
@@ -187,8 +184,7 @@ func (s *InitialSync) roundRobinSync(genesis time.Time) error {
 	// mitigation. We are already convinced that we are on the correct finalized chain. Any blocks
 	// we receive there after must build on the finalized chain or be considered invalid during
 	// fork choice resolution / block processing.
-	best := bestPeer()
```
Comment (prestonvanloon) [inline beacon-chain/sync/initial-sync/round_robin.go]: Why change this?

File: beacon-chain/sync/initial-sync/round_robin.go (hunk 4 above)
Comment (nisdas) [reply beacon-chain/sync/initial-sync/round_robin.go]: +1

File: beacon-chain/sync/initial-sync/round_robin.go (hunk 2 above)
Comment (nisdas) [reply beacon-chain/sync/initial-sync/round_robin.go]: Can we change back to the old names ? The 3 letter vars arent clear enough

File: beacon-chain/sync/initial-sync/round_robin.go (hunk 5 above)
Comment (shayzluf) [reply beacon-chain/sync/initial-sync/round_robin.go]: i had a case where the chosen peer had failure on these slots. then it all fails

File: beacon-chain/sync/initial-sync/round_robin.go (hunk 4 above)
Comment (shayzluf) [reply beacon-chain/sync/initial-sync/round_robin.go]: ```DEBU[0003] helpers.StartSlot(highestFinalizedEpoch()+1)-1: 127  head status:=124 prefix=initial-sync```

Comment (shayzluf) [reply beacon-chain/sync/initial-sync/round_robin.go]:
```DEBU[0003] Requesting blocks...                          peer=12D3KooWE4qDcRrueTuRYWUdQZgcy7APZqBngVeXRt4Y6ytHizKV prefix=initial-sync req="head_block_root:\"finalized_root 1\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\" start_slot:114557 count:64 step:4 "
```

Comment (shayzluf) [reply beacon-chain/sync/initial-sync/round_robin.go]: it seems strange as this is the code that calls it ```blocks, err := request(
			s.chain.HeadSlot()+1, // start
			1,                    // step
			blockBatchSize,       // count
			peers,                // peers
			0,                    // remainder
		)```

File: beacon-chain/sync/initial-sync/round_robin.go (hunk 5 above)
Comment (shayzluf) [reply beacon-chain/sync/initial-sync/round_robin.go]: better not to be dependent on a single node

File: beacon-chain/sync/initial-sync/round_robin.go (hunk 3 above)
Comment (shayzluf) [reply beacon-chain/sync/initial-sync/round_robin.go]: fixed

File: beacon-chain/sync/initial-sync/round_robin.go (hunk 4 above)
Comment (shayzluf) [reply beacon-chain/sync/initial-sync/round_robin.go]: found the issue + fixed

**Pull request closed without merging**